File: cmd/dependabot-directories/detect_test.go
package main

// Ports of the bats detection tests, one for one, driving the real CLI over
// the tree fixtures in --detect-only mode.

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// --- npm detection ----------------------------------------------------------

func TestRootNpmRepoMapsTheRootOnce(t *testing.T) {
	code, out := detectFixture(t, "root-npm")
	assertExit(t, code, 0, out)
	assertMapping(t, out, "npm: /")
	assertMapping(t, out, "github-actions: /")
	if n := mappingCount(out); n != 2 {
		t.Errorf("mapping count = %d, want 2\noutput:\n%s", n, out)
	}
}

func TestYarnWorkspacesMapOnlyTheRootNotEveryPackage(t *testing.T) {
	code, out := detectFixture(t, "yarn-workspaces")
	assertExit(t, code, 0, out)
	assertMapping(t, out, "npm: /")
	// The rule that prevents per-package entry spam in hoisted monorepos.
	refuteMapping(t, out, "/packages/a")
}

func TestAWorkspaceCoveredPackageWithItsOwnLockfileIsReportedNotMapped(t *testing.T) {
	_, out := detectFixture(t, "yarn-workspaces")
	refuteMapping(t, out, "/packages/b")
	assertContains(t, out, "shadowed by the workspace")
	assertNoted(t, out, "/packages/b")
}

func TestAnUncoveredManifestWithNoLockfileIsReportedNotMapped(t *testing.T) {
	_, out := detectFixture(t, "yarn-workspaces")
	refuteMapping(t, out, "/tools/helper")
	assertContains(t, out, "no lockfile and no workspace covering them")
	assertNoted(t, out, "/tools/helper")
}

func TestYarnV1ObjectFormWorkspacesAreParsed(t *testing.T) {
	code, out := detectFixture(t, "workspaces-object-form")
	assertExit(t, code, 0, out)
	assertMapping(t, out, "npm: /")
	if n := mappingCount(out); n != 1 {
		t.Errorf("mapping count = %d, want 1\noutput:\n%s", n, out)
	}
	// Parsed means covered: neither package may surface as an entry or a
	// finding.
	refuteNoted(t, out, "/packages/a")
	refuteNoted(t, out, "/packages/b")
}

func TestANegatedWorkspacePatternLeavesThatPackageToFendForItself(t *testing.T) {
	code, out := detectFixture(t, "workspaces-negation")
	assertExit(t, code, 0, out)
	assertMapping(t, out, "npm: /")
	// Excluded from the workspace and holding its own lockfile, so it
	// needs an entry.
	assertMapping(t, out, "npm: /packages/legacy")
}

func TestGlobstarCoversTheZeroSegmentCase(t *testing.T) {
	code, out := detectFixture(t, "workspaces-globstar")
	assertExit(t, code, 0, out)
	assertMapping(t, out, "npm: /")
	// apps/**/web must match apps/web as well as apps/deep/nested/web.
	refuteNoted(t, out, "/apps/web")
	refuteNoted(t, out, "/apps/deep/nested/web")
	// Outside the pattern, so it is still reported.
	assertNoted(t, out, "/other/thing")
}

func TestAnUnrecognisedWorkspacesShapeIsReportedRatherThanAssumedAbsent(t *testing.T) {
	code, out := detectFixture(t, "workspaces-unknown-shape")
	assertExit(t, code, 0, out)
	assertContains(t, out, "not a shape we recognise")
	assertMapping(t, out, "npm: /")
}

func TestPnpmWorkspaceYamlPackagesAreHonoured(t *testing.T) {
	code, out := detectFixture(t, "pnpm")
	assertExit(t, code, 0, out)
	assertMapping(t, out, "npm: /")
	if n := mappingCount(out); n != 1 {
		t.Errorf("mapping count = %d, want 1\noutput:\n%s", n, out)
	}
}

func TestAPnpmWorkspaceWithNoPackagesKeyIsReported(t *testing.T) {
	code, out := detectFixture(t, "pnpm-no-packages-key")
	assertExit(t, code, 0, out)
	assertContains(t, out, "declares no 'packages:'")
	assertMapping(t, out, "npm: /")
}

// --- composer detection -----------------------------------------------------

func TestComposerNeedsBothAManifestAndALock(t *testing.T) {
	code, out := detectFixture(t, "composer-single")
	assertExit(t, code, 0, out)
	assertMapping(t, out, "composer: /")
}

// --- exclusions -------------------------------------------------------------

func TestHardExcludedDirectoriesAreDroppedSilently(t *testing.T) {
	code, out := detectFixture(t, "excluded")
	assertExit(t, code, 0, out)
	refuteMapping(t, out, "node_modules")
	refuteMapping(t, out, "vendor")
	if strings.Contains(out, "node_modules") {
		t.Errorf("hard exclusions must be silent\noutput:\n%s", out)
	}
}

func TestSoftExcludedDirectoriesAreDroppedWithANote(t *testing.T) {
	_, out := detectFixture(t, "excluded")
	refuteMapping(t, out, "examples")
	assertContains(t, out, "soft-excluded")
}

func TestIncludeReadmitsASoftExcludedDirectory(t *testing.T) {
	code, out := detectFixture(t, "excluded", "--include", "examples")
	assertExit(t, code, 0, out)
	assertMapping(t, out, "composer: /examples/demo")
}

func TestIncludeEqualsFormIsAccepted(t *testing.T) {
	code, out := detectFixture(t, "excluded", "--include=examples")
	assertExit(t, code, 0, out)
}

// --- deferred ecosystems ----------------------------------------------------

func TestDeferredEcosystemLockfilesAreReported(t *testing.T) {
	code, out := detectFixture(t, "deferred-ecosystems")
	assertExit(t, code, 0, out)
	for _, lock := range []string{"Gemfile.lock", "go.sum", "Cargo.lock", "poetry.lock", "uv.lock"} {
		assertContains(t, out, lock)
	}
	assertMapping(t, out, "composer: /")
}

// --- mixed and empty --------------------------------------------------------

func TestARepoWithNoManifestsReportsAndExitsClean(t *testing.T) {
	code, out := detectFixture(t, "no-manifests")
	assertExit(t, code, 0, out)
	assertContains(t, out, "no npm, composer or github-actions manifests")
}

func TestWorkflowsAloneMapGithubActionsAtTheRoot(t *testing.T) {
	code, out := detectFixture(t, "actions-only")
	assertExit(t, code, 0, out)
	assertMapping(t, out, "github-actions: /")
	if n := mappingCount(out); n != 1 {
		t.Errorf("mapping count = %d, want 1\noutput:\n%s", n, out)
	}
}

// Ports of the bats grouping, guard, awkward-path and sanity-cap tests.

func TestSiblingComposerPackagesCollapseIntoOneGlob(t *testing.T) {
	code, out := detectFixture(t, "composer-siblings")
	assertExit(t, code, 0, out)
	assertMapping(t, out, "composer: /projects/plugins/* (glob)")
	if n := mappingCount(out); n != 1 {
		t.Errorf("mapping count = %d, want 1\noutput:\n%s", n, out)
	}
}

func TestGuard2AGlobThatWouldSweepAnUnmappedSiblingDegradesToSingulars(t *testing.T) {
	code, out := detectFixture(t, "composer-siblings-lockless")
	assertExit(t, code, 0, out)
	refuteMapping(t, out, "*")
	assertMapping(t, out, "composer: /projects/plugins/a")
	assertMapping(t, out, "composer: /projects/plugins/b")
	assertMapping(t, out, "composer: /projects/plugins/c")
	assertContains(t, out, "did not map")
	// d has a manifest but no lock, so it is reported and never mapped.
	refuteMapping(t, out, "/projects/plugins/d")
}

func TestGuard1TopLevelSiblingsNeverProduceARootGlob(t *testing.T) {
	code, out := detectFixture(t, "top-level-siblings")
	assertExit(t, code, 0, out)
	refuteMapping(t, out, "*")
	assertMapping(t, out, "composer: /a")
	assertMapping(t, out, "composer: /b")
	assertContains(t, out, "repository root")
}

// --- awkward paths ----------------------------------------------------------

func TestPathsWithRegexMetacharactersAndUnicodeSurviveTranslation(t *testing.T) {
	code, out := detectFixture(t, "weird-paths")
	assertExit(t, code, 0, out)
	assertMapping(t, out, "composer: /my dir")
	assertMapping(t, out, "composer: /p+kg")
	assertMapping(t, out, "composer: /br[ack]et")
	assertMapping(t, out, "composer: /uni_café")
}

func TestAPathContainingAPipeIsProcessedLikeAnyOther(t *testing.T) {
	// The bash script skipped these paths to protect its '|'-delimited
	// record format and said so with a note. The record format is gone, so
	// the pipe directory flows through detection like any other: it holds
	// a composer.json with no lock, which lands it in that note group.
	_, out := detectFixture(t, "weird-paths")
	if strings.Contains(out, "containing '|'") {
		t.Errorf("the record-format skip note must be gone\noutput:\n%s", out)
	}
	refuteMapping(t, out, "bad|pipe")
	assertNoted(t, out, "/bad|pipe")
}

// --- mixed ------------------------------------------------------------------

func TestAMixedMonorepoMapsThePnpmRootAndGlobsTheComposerPlugins(t *testing.T) {
	code, out := detectFixture(t, "mixed-monorepo")
	assertExit(t, code, 0, out)
	assertMapping(t, out, "npm: /")
	assertMapping(t, out, "composer: /projects/plugins/* (glob)")
	assertMapping(t, out, "github-actions: /")
}

// --- sanity cap -------------------------------------------------------------

func TestAnImplausibleNumberOfMappingsStopsTheRunUnlessForced(t *testing.T) {
	var tree strings.Builder
	for i := 1; i <= 60; i++ {
		fmt.Fprintf(&tree, "pkg%d/composer.json\npkg%d/composer.lock\n", i, i)
	}
	pathsFile := filepath.Join(t.TempDir(), "many.paths")
	if err := os.WriteFile(pathsFile, []byte(tree.String()), 0o644); err != nil {
		t.Fatal(err)
	}

	code, out := execute(t, "acme/widgets", "--paths-from-file", pathsFile, "--detect-only")
	assertExit(t, code, 1, out)
	assertContains(t, out, "sanity cap")

	code, out = execute(t, "acme/widgets", "--paths-from-file", pathsFile, "--detect-only", "--force")
	assertExit(t, code, 0, out)
}

File: tests/Feature/DetectTest.php
<?php

declare(strict_types=1);

/*
 * Ports of the bats detection tests, one for one, driving the real CLI over
 * the tree fixtures in --detect-only mode.
 */

// --- npm detection ----------------------------------------------------------

it('maps the root once for a root npm repo', function () {
    [$code, $out] = detectFixture('root-npm');

    expect($code)->toBe(0);
    expect($out)->toListMapping('npm: /')
        ->toListMapping('github-actions: /')
        ->toListMappingCount(2);
});

it('maps only the workspace root, not every package', function () {
    [$code, $out] = detectFixture('yarn-workspaces');

    expect($code)->toBe(0);
    expect($out)->toListMapping('npm: /')
        // The rule that prevents per-package entry spam in hoisted monorepos.
        ->toNotListMapping('/packages/a');
});

it('reports a workspace-covered package with its own lockfile rather than mapping it', function () {
    [, $out] = detectFixture('yarn-workspaces');

    expect($out)->toNotListMapping('/packages/b')
        ->toContain('shadowed by the workspace')
        ->toNoteExample('/packages/b');
});

it('reports an uncovered manifest with no lockfile rather than mapping it', function () {
    [, $out] = detectFixture('yarn-workspaces');

    expect($out)->toNotListMapping('/tools/helper')
        ->toContain('no lockfile and no workspace covering them')
        ->toNoteExample('/tools/helper');
});

it('parses the yarn v1 object form of workspaces', function () {
    [$code, $out] = detectFixture('workspaces-object-form');

    expect($code)->toBe(0);
    expect($out)->toListMapping('npm: /')->toListMappingCount(1)
        // Parsed means covered: neither package may surface as an entry or a
        // finding.
        ->toNotNoteExample('/packages/a')
        ->toNotNoteExample('/packages/b');
});

it('leaves a negated workspace package to fend for itself', function () {
    [$code, $out] = detectFixture('workspaces-negation');

    expect($code)->toBe(0);
    expect($out)->toListMapping('npm: /')
        // Excluded from the workspace and holding its own lockfile, so it
        // needs an entry.
        ->toListMapping('npm: /packages/legacy');
});

it('covers the zero-segment globstar case', function () {
    [$code, $out] = detectFixture('workspaces-globstar');

    expect($code)->toBe(0);
    expect($out)->toListMapping('npm: /')
        // apps/**/web must match apps/web as well as apps/deep/nested/web.
        ->toNotNoteExample('/apps/web')
        ->toNotNoteExample('/apps/deep/nested/web')
        // Outside the pattern, so it is still reported.
        ->toNoteExample('/other/thing');
});

it('reports an unrecognised workspaces shape rather than assuming it absent', function () {
    [$code, $out] = detectFixture('workspaces-unknown-shape');

    expect($code)->toBe(0);
    expect($out)->toContain('not a shape we recognise')->toListMapping('npm: /');
});

it('honours pnpm-workspace.yaml packages', function () {
    [$code, $out] = detectFixture('pnpm');

    expect($code)->toBe(0);
    expect($out)->toListMapping('npm: /')->toListMappingCount(1);
});

it('reports a pnpm workspace with no packages key', function () {
    [$code, $out] = detectFixture('pnpm-no-packages-key');

    expect($code)->toBe(0);
    expect($out)->toContain("declares no 'packages:'")->toListMapping('npm: /');
});

// --- composer detection -----------------------------------------------------

it('needs both a composer manifest and a lock', function () {
    [$code, $out] = detectFixture('composer-single');

    expect($code)->toBe(0);
    expect($out)->toListMapping('composer: /');
});

// --- exclusions -------------------------------------------------------------

it('drops hard-excluded directories silently', function () {
    [$code, $out] = detectFixture('excluded');

    expect($code)->toBe(0);
    expect($out)->toNotListMapping('node_modules')
        ->toNotListMapping('vendor')
        // Silent means silent: not even a note.
        ->not->toContain('node_modules');
});

it('drops soft-excluded directories with a note', function () {
    [, $out] = detectFixture('excluded');

    expect($out)->toNotListMapping('examples')->toContain('soft-excluded');
});

it('readmits a soft-excluded directory with --include', function () {
    [$code, $out] = detectFixture('excluded', ['--include', 'examples']);

    expect($code)->toBe(0);
    expect($out)->toListMapping('composer: /examples/demo');
});

it('accepts the --include=value form', function () {
    [$code] = detectFixture('excluded', ['--include=examples']);

    expect($code)->toBe(0);
});

// --- deferred ecosystems ----------------------------------------------------

it('reports deferred-ecosystem lockfiles', function () {
    [$code, $out] = detectFixture('deferred-ecosystems');

    expect($code)->toBe(0);
    foreach (['Gemfile.lock', 'go.sum', 'Cargo.lock', 'poetry.lock', 'uv.lock'] as $lockfile) {
        expect($out)->toContain($lockfile);
    }
    expect($out)->toListMapping('composer: /');
});

// --- mixed and empty --------------------------------------------------------

it('reports and exits clean for a repo with no manifests', function () {
    [$code, $out] = detectFixture('no-manifests');

    expect($code)->toBe(0);
    expect($out)->toContain('no npm, composer or github-actions manifests');
});

it('maps github-actions at the root for workflows alone', function () {
    [$code, $out] = detectFixture('actions-only');

    expect($code)->toBe(0);
    expect($out)->toListMapping('github-actions: /')->toListMappingCount(1);
});

// --- grouping and guards ----------------------------------------------------

it('collapses sibling composer packages into one glob', function () {
    [$code, $out] = detectFixture('composer-siblings');

    expect($code)->toBe(0);
    expect($out)->toListMapping('composer: /projects/plugins/* (glob)')->toListMappingCount(1);
});

it('degrades to singulars when a glob would sweep an unmapped sibling', function () {
    [$code, $out] = detectFixture('composer-siblings-lockless');

    expect($code)->toBe(0);
    expect($out)->toNotListMapping('*')
        ->toListMapping('composer: /projects/plugins/a')
        ->toListMapping('composer: /projects/plugins/b')
        ->toListMapping('composer: /projects/plugins/c')
        ->toContain('did not map')
        // d has a manifest but no lock, so it is reported and never mapped.
        ->toNotListMapping('/projects/plugins/d');
});

it('never produces a root glob for top-level siblings', function () {
    [$code, $out] = detectFixture('top-level-siblings');

    expect($code)->toBe(0);
    expect($out)->toNotListMapping('*')
        ->toListMapping('composer: /a')
        ->toListMapping('composer: /b')
        ->toContain('repository root');
});

// --- awkward paths ----------------------------------------------------------

it('survives regex metacharacters and unicode in paths', function () {
    [$code, $out] = detectFixture('weird-paths');

    expect($code)->toBe(0);
    expect($out)->toListMapping('composer: /my dir')
        ->toListMapping('composer: /p+kg')
        ->toListMapping('composer: /br[ack]et')
        ->toListMapping('composer: /uni_café');
});

it('processes a path containing a pipe like any other', function () {
    // The bash script skipped these paths to protect its '|'-delimited record
    // format and said so with a note. The record format is gone, so the pipe
    // directory flows through detection like any other: it holds a
    // composer.json with no lock, which lands it in that note group.
    [, $out] = detectFixture('weird-paths');

    expect($out)->not->toContain("containing '|'");
    expect($out)->toNotListMapping('bad|pipe')->toNoteExample('/bad|pipe');
});

// --- mixed ------------------------------------------------------------------

it('maps the pnpm root and globs the composer plugins in a mixed monorepo', function () {
    [$code, $out] = detectFixture('mixed-monorepo');

    expect($code)->toBe(0);
    expect($out)->toListMapping('npm: /')
        ->toListMapping('composer: /projects/plugins/* (glob)')
        ->toListMapping('github-actions: /');
});

// --- sanity cap -------------------------------------------------------------

it('stops an implausible number of mappings unless forced', function () {
    $tree = '';
    for ($i = 1; $i <= 60; ++$i) {
        $tree .= "pkg{$i}/composer.json\npkg{$i}/composer.lock\n";
    }
    $pathsFile = sys_get_temp_dir().'/many-'.bin2hex(random_bytes(6)).'.paths';
    file_put_contents($pathsFile, $tree);

    [$code, $out] = runTool(['acme/widgets', '--paths-from-file', $pathsFile, '--detect-only']);
    expect($code)->toBe(1);
    expect($out)->toContain('sanity cap');

    [$code] = runTool(['acme/widgets', '--paths-from-file', $pathsFile, '--detect-only', '--force']);
    expect($code)->toBe(0);

    unlink($pathsFile);
});
